src/java.base/share/classes/jdk/internal/foreign/abi/s390/linux/LinuxS390CallArranger.java

```diff
@@ -0,0 +1,312 @@
+
+            short encodedSize = (short) size;
+            assert (encodedSize & 0xFFFF) == size;
+
+            VMStorage storage = stackStorage(encodedSize, (int) alignedStackOffset);
+            stackOffset = alignedStackOffset + size;
+            return storage;
+        }
+
+        VMStorage regAlloc(int type) {
+            int gpRegCnt = (type == StorageType.INTEGER) ? 1 : 0;
+            int fpRegCnt = (type == StorageType.FLOAT) ? 1 : 0;
+
+            // Use stack if not enough registers available.
+            if ((type == StorageType.FLOAT && (nRegs[StorageType.FLOAT] + fpRegCnt) > MAX_FLOAT_REGISTER_ARGUMENTS)
+                    || (type == StorageType.INTEGER && (nRegs[StorageType.INTEGER] + gpRegCnt) > MAX_REGISTER_ARGUMENTS)) return null;
+
+            VMStorage[] source = (forArguments ? CLinux.inputStorage : CLinux.outputStorage)[type];
+            VMStorage result = source[nRegs[type]];
+
+            nRegs[StorageType.INTEGER] += gpRegCnt;
+            nRegs[StorageType.FLOAT] += fpRegCnt;
+            return result;
+
+        }
+        VMStorage getStorage(int type, boolean is32Bit) {
+            VMStorage reg = regAlloc(type);
+            if (reg != null) {
+                if (is32Bit) {
+                    reg = new VMStorage(reg.type(), REG32_MASK, reg.indexOrOffset());
+                }
+                return reg;
+            }
+            VMStorage stack;
+            if (is32Bit) {
+                stackAlloc(4, STACK_SLOT_SIZE); // Skip first half of stack slot.
+                stack = stackAlloc(4, 4);
+            } else
+                stack = stackAlloc(8, STACK_SLOT_SIZE);
+
+            return stack;
+        }
+    }
+
+    abstract static class BindingCalculator {
+        protected final StorageCalculator storageCalculator;
+
+        protected BindingCalculator(boolean forArguments) {
+            this.storageCalculator = new LinuxS390CallArranger.StorageCalculator(forArguments);
+        }
+
+        abstract List<Binding> getBindings(Class<?> carrier, MemoryLayout layout);
+    }
+
+    // Compute recipe for transferring arguments / return values to C from Java.
+    static class UnboxBindingCalculator extends BindingCalculator {
+        UnboxBindingCalculator(boolean forArguments) {
+            super(forArguments);
+        }
+
+        @Override
+        List<Binding> getBindings(Class<?> carrier, MemoryLayout layout) {
+            TypeClass argumentClass = TypeClass.classifyLayout(layout);
+            Binding.Builder bindings = Binding.builder();
+            switch (argumentClass) {
+                case STRUCT_REGISTER -> {
+                    assert carrier == MemorySegment.class;
+                    VMStorage storage = storageCalculator.getStorage(StorageType.INTEGER, false);
+                    Class<?> type = SharedUtils.primitiveCarrierForSize(layout.byteSize(), false);
+                    bindings.bufferLoad(0, type)
+                        .vmStore(storage, type);
```

Comment:
Maybe improve indentation?

Reply:
Thank you @TheRealMDoerr. Fixed in new commit.